File: LeetCode/Problems/Python/953. Verifying an Alien Dictionary.py
#Time: O(C), C is the length of all the words
#Space: O(1), because order bounded by 26
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        if not words or not order:
            return False
        
        aliendict = {}
        for idx, ch in enumerate(order):
            aliendict[ch] = idx
            
        #extract relationship
        for w1, w2 in zip(words, words[1:]):
            for ch in range(min(len(w1), len(w2))):
                if w1[ch] != w2[ch]:
                    if aliendict[w1[ch]] > aliendict[w2[ch]]:
                        return False
                    break
            else:
                if len(w2) < len(w1):
                    return False
        
        return True

File: LeetCode/Problems/C++/953. Verifying an Alien Dictionary.cpp
class Solution {
public:
    bool isAlienSorted(vector<string>& words, string order) {
        if(!words.size()) return false;
        if(words.size()==1) return true;
        
        unordered_map<char,int> lookup;
        for(int i=0;i<order.size();i++){
            lookup[order[i]] = i;
        }
        
        for(int i=0;i<words.size()-1;i++){
            string s1 = words[i];
            string s2 = words[i+1];
            int p1 = 0;
            int p2 = 0;
            while(p1<s1.size() && p2<s2.size()){
                if(s1[p1]!=s2[p2]){
                    if(lookup[s1[p1]] > lookup[s2[p2]]) return false;
                    break;
                }
                p1++;
                p2++;  
            }
            if(s1.substr(0,p1)==s2 && p1 < s1.size()) return false;
        }
        
        return true;
    }
};